File: Pragma_Examples/OpenMP/Fortran/3_vecadd/5_vecadd_async/vecadd.f90
! Copyright (c) 2022 Advanced Micro Devices, Inc. All rights reserved.
! 
program main

    use omp_lib 
    ! Size of vectors
    integer :: n = 10000000
 
    ! Input vectors and Output vector
    real(8),dimension(:),allocatable :: a, b, c
 
    integer :: i
    real(8) :: sum
    real(8) :: startt, endt
 
    ! Allocate memory for each vector
    allocate(a(n), b(n), c(n))
 
    !$omp target enter data map(alloc:a,b,c)
    ! Initialize input vectors.
    !$omp target teams distribute parallel do simd nowait depend(out:a)
    do i=1,n
        a(i) = sin(dble(i)*1.0d0)*sin(dble(i)*1.0d0)
    enddo
    !$omp end target teams distribute parallel do simd

    !$omp target teams distribute parallel do simd nowait depend(out:b)
    do i=1,n
        b(i) = cos(dble(i)*1.0d0)*cos(dble(i)*1.0d0) 
    enddo
    !$omp end target teams distribute parallel do simd
    !$omp target teams distribute parallel do simd nowait depend(out:c)
    do i=1,n
        c(i) = 0.0d0
    enddo
    !$omp end target teams distribute parallel do simd
    !meassure after warmup kernel
    startt=omp_get_wtime()
    ! Sum each component of arrays

    !$omp target teams distribute parallel do simd nowait depend(in:a,b,c) depend(out:c)
    do i=1,n
        c(i) = a(i) + b(i)
    enddo
    !$omp end target teams distribute parallel do simd

    ! Sum up vector c. Print result divided by n. It should equal 1
    sum = 0.0d0
    !$omp target teams distribute parallel do simd reduction(+:sum) nowait depend(in:c)
    do i=1,n
        sum = sum +  c(i)
    enddo
    !$omp end target teams distribute parallel do simd
    !$omp taskwait
    !$omp target exit data map(release:a,b,c)

    sum = sum/dble(n)
    write(*,'("Final result: ",f10.6)') sum

    endt=omp_get_wtime()
    write(*,'("Runtime is: ",f8.6," secs")') endt-startt
 
    ! Deallocate memory
    deallocate(a, b, c)
 
end program

File: Pragma_Examples/OpenMP/C/3_vecadd/3_vecadd_targetdata/vecadd.c
/*
Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.

*/
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#include <math.h>


int main()
{
  int N = 100000;

  // Input arrays
  double a[N];
  double b[N];
  // Output arrays
  double c[N];

  double sum;
  double tstart,tend;
  #pragma omp target data map(alloc:a[0:N],b[0:N],c[0:N]) map(from:sum,tstart,tend)
  {
  #pragma omp target teams distribute parallel for
  for (int i=0; i<N; i++){
    a[i]=sin((double)(i+1))*sin((double)(i+1));
    b[i]=cos((double)(i+1))*cos((double)(i+1));
    c[i]=0.0;
  }
  tstart = omp_get_wtime();
  #pragma omp target teams distribute parallel for
  for (int j = 0; j< N; j++){
    c[j] = a[j] + b[j];
  }

  sum = 0.0;
  #pragma omp target teams distribute parallel for reduction(+: sum)
  for (int i=0; i<N; i++){
    sum += c[i];
  }

  tend = omp_get_wtime();
  }
  sum = sum/(double)N;

  printf("Runtime is: %lf secs\n", tend-tstart);

  printf("Final result: %lf\n",sum);

}
